File: src/com/aslan/module/notify/HashClassifyNotification.js
import ClassifyNotification from "./ClassifyNotification";

let statics = new Map();

export default class HashClassifyNotification extends ClassifyNotification {

    content = new Map();

    addNotifier(name, notifier, global) {
        let map = global ? statics : this.content;
        let list = map.get(name);
        if (!list) {
            map.set(name, list = []);
        }
        list.push(notifier);
    }

    removeNotifier(name, notifier, global) {
        let map = global ? statics : this.content;
        let list = map.get(name);
        if (list) {
            list.remove(notifier);
        }
    }

    notify(event) {
        this.doNotify(event, statics.get(event.name));
        if (!event.stopPropagation) {
            this.doNotify(event, this.content.get(event.name));
        }
    }

    doNotify(event, list) {
        let size = (list && list.length) || 0;
        if (size === 0) return;
        for (let i = 0; i < size; i++) {
            try {
                let notifier = list[i];
                if (notifier instanceof Function) {
                    notifier.call(this, event);
                } else {
                    list[i].action(event);
                }
                if (event.stopImmediatePropagation) break;
            } catch (error) {
                console.error(error);
            }
        }
    }
}

File: src/com/aslan/module/notify/HashClassifyNotification.java
package com.aslan.module.notify;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class HashClassifyNotification implements ClassifyNotification {
    Map<String, List<Notifier>> statics = new HashMap();
    Map<String, List<Notifier>> content = new HashMap<>();

    public void addNotifier(String name, Notifier notifier, boolean global) {
        Map<String, List<Notifier>> map = global ? statics : this.content;
        List<Notifier> list = map.computeIfAbsent(name, (key) -> new ArrayList<>());
        list.add(notifier);
    }

    public void removeNotifier(String name, Notifier notifier, boolean global) {
        Map<String, List<Notifier>> map = global ? statics : this.content;
        List<Notifier> list = map.get(name);
        if (list != null) {
            list.remove(notifier);
        }
    }

    public void notify(NotifyEvent event) {
        doNotify(event, statics.get(event.getName()));
        if (!event.isStopPropagation()) {
            doNotify(event, this.content.get(event.getName()));
        }
    }

    protected void doNotify(NotifyEvent event, List<Notifier> list) {
        int size = list == null ? 0 : list.size();
        for (int i = 0; i < size; i++) {
            try {
                list.get(i).action(event);
                if (event.isStopImmediatePropagation()) break;
            } catch (Exception e) {

            }
        }
    }
}